### PR TITLE
Added a separate bed timeout enable

### PR DESCRIPTION
**Pull request by MachX428**
I added a new setting/checkbox to control the Bed Heater timeout separately from the Extruder timeout. Using the two "enable" settings will allow a user the option of the plugin shutting off the Extruder, the Bed, or both depending on their needs.  I personally needed this so that the plugin didn't shut off my Bed if a print paused for a filament change.

The plugin version in this PR may be installed using URL:
https://github.com/MachX428/OctoPrint-BetterHeaterTimeout/archive/BedEnable.zip

### CODE REVIEW

**Comment by MachX428** [comment]
One more change. I noticed that the settings page didn't make much sense anymore with the Hot End just labeled "Timeout" and the setting for from when the timeout is triggered seemingly just associated with the hotend. I renamed "Timeout" to "Hot End" and moved the other setting to the bottom of that group. I then squashed it all up into one commit for a cleaner log.

**Comment by MachX428** [comment]
Make that, renamed it to "Tool" instead of "Hot End", to follow the OctoPrint naming convention.

**Comment by tjjfvi** [comment]
Changes look good, but I can't easily test atm. Is this tested on your end? Also, could I get a quick screenshot of the new settings page?

**Comment by MachX428** [comment]
Yes, I tested it yesterday, and I have had it running all day. Since you aren't personally able to validate at the moment, I spent sometime putting together a quick test plan and doing a thorough run down. 

**TLDR, it all works as expected** - HOWEVER while testing I found a different issue that exists even in the currently released code (ie, has nothing to do with my change). It seems that when you have the tool and bed triggers set to the same timeout, when they trigger, you actually get 3 events instead of 2 -  1 x TOOL, 2 x BED. This is probably not a huge deal to most people, but I could see if someone was doing something special in the gcodes sections it could cause an issue. It seems to be a race condition caused by the gcodes (including the set target temp to 0) queueing up and the plugin getting another callback before the bed as actually has reset its target temperature. **Anyway, I have a fix for this that I am testing and can include it in the PR (or a different PR) if you want as well.**

Here is the screenshot you asked for: 
BetterHeaterTimeout-Settings


For completeness, here are the testing steps I went through and confirmed to verify this PR works as expected:

### Test Steps:

**Validate that the new version can install correctly, and set up for further testing**

- Uninstall the original BetterHeaterTimeout plugin using Plugin Mananger
  - Choose Uninstall and Cleanup Data
- Restart Octoprint
- Install new version of plugin using Plugin Manager -> Get More -> Install from URL
  - Use this link: [https://github.com/MachX428/OctoPrint-BetterHeaterTimeout/archive/BedEnable.zip](https://github.com/MachX428/OctoPrint-BetterHeaterTimeout/archive/BedEnable.zip)   
- Go to OctoPrint settings
  - Make sure BetterHeaterTimout Appears under Plugins
  - Select it and go to the settings page
  
**Test both Tool and Bed Trigger appropriately when both enabled with same times**
  - Set Tool and Bed to 30
  - Make sure both Tool and Bed are enabled
  - Set From to "after target temp changes"
  - Save
  - Go to Temperature->Target->Set PLA (Tool: 180C / Bed: 60C)
  - Temperature for both Tool and Bed should start increasing
  - 30 seconds later, you should get the following messages
    - "HeaterTimeout - Heater 'tool0' disabled after 31 seconds."
    - "HeaterTimeout - Heater 'bed' disabled after 31 seconds."
    - _NOTE -  I also got a second Heater 'bed' timeout in this  case every time. See my note above. Also, I'm not sure why it always gives me 31 seconds when I am asking for 30._
  - You should visually see target temperature go to 0 for both and the tool and bed temperature start decreasing

**Test both Tool and Bed Trigger appropriately when both enabled with different times**
  - Set Tool to 30 and enabled
  - Set Bed to 60 and enabled
  - Set From to "after target temp changes"
  - Save
  - Go to Temperature->Target->Set PLA (Tool: 180C / Bed: 60C)
  - Temperature for both Tool and Bed should start increasing
  - 30 seconds later, you should get the following message
    - HeaterTimeout - Heater 'tool0' disabled after 31 seconds."
    - You should visually see target temperature go to 0 for the tool and the temperature start decreasing
  - after another 30 seconds, you should get the following message
    - "HeaterTimeout - Heater 'bed' disabled after 61 seconds."
    - You should visually see target temperature go to 0 for the bed and the temperature start decreasing

**Test only Bed Triggers when only Bed enabled**
  - Go to BetterHeaterTimeout Settings
  - Set Tool to 30 and DISABLED
  - Set Bed to 30 and ENABLED
  - Set From to "after target temp changes"
  - Save
  - Go to Temperature->Target->Set PLA (Tool: 180C / Bed: 60C)
  - Temperature for both Tool and Bed should start increasing
  - 30 seconds later, you should get the following messages
    - "HeaterTimeout - Heater 'bed' disabled after 31 seconds."
    - You should visually see the tool target temperature go to 0 and the tool temperature start decreasing
    - The target temperature of the bed should remain the same, and the bed temperature should not start decreasing

**Test only Tool Triggers when only Tool enabled**
  - Go to BetterHeaterTimeout Settings
  - Set Tool to 30 and ENABLED
  - Set Bed to 30 and DISABLED
  - Set From to "after target temp changes"
  - Save
  - Go to Temperature->Target->Set PLA (Tool: 180C / Bed: 60C)
  - Temperature for both Tool and Bed should start increasing
  - 30 seconds later, you should get the following messages
    - "HeaterTimeout - Heater 'tool0' disabled after 31 seconds."
    - You should visually see the bed target temperature go to 0 and the bed temperature start decreasing
    - The target temperature of the tool should remain the same, and the tool temperature should not start decreasing

**Test both Tool and Bed Trigger appropriately when from is set to _"after heating starts"_**
  - Set Tool to 30 and enabled
  - Set Bed to 60 and enabled
  - Set From to "after heating starts"
  - Save
  - Go to Temperature->Target->Set PLA (Tool: 180C / Bed: 60C)
  - Temperature for both Tool and Bed should start increasing
  - 30 seconds later, you should get the following message
    - "HeaterTimeout - Heater 'tool0' disabled after 31 seconds."
    - You should visually see target temperature go to 0 for the tool and the temperature start decreasing
  - after another 30 seconds, you should get the following message
    - "HeaterTimeout - Heater 'bed' disabled after 61 seconds."
    - You should visually see target temperature go to 0 for the bed and the temperature start decreasing

**Test that event is not triggered if heater is not on**
  - Go to BetterHeaterTimeout settings 
  - Set Tool and Bed to 30
  - Make sure both Tool and Bed are enabled
  - Set From to "after target temp changes"
  - Save
  - Go to Temperature->Tool->Set PLA (Tool: 180C)
    - Do not set Bed Temperature target 
  - Temperature for Tool  should start increasing
  - 30 seconds later, you should get the following message
    - "HeaterTimeout - Heater 'tool0' disabled after 31 seconds."
    - You should visually see target temperature go to 0 for the tool and the temperature start decreasing
    - You should not get a message for the bed.

**Comment by MachX428** [comment]
Actually, that other bug I mentioned above that I found while testing is almost certainly the same thing that is causing issue #17 . The fix I am looking (resetting the timer in self._temp_statuses[key] when the timeout is triggered) solves the issue I am seeing with the 1 extra timeout. That said, it would not necessarily prevent multiple triggers if you have something taking longer than double your idle timeout before processing the rest of the gcode queue like his 30 minute bed level. But - it would no longer constantly spam. In his particular case he would likely get 4 timeouts (2 each for the bed and tool) instead of thousands. Just let me know if you want me to check in the fix - either as part of this or as a different PR.

**Comment by tjjfvi** [comment]
@MachX428 ...wow. Thanks for such thorough testing! I don't think I've explicitly tested this plugin that extensively myself 😅 

Settings page looks good 👍 

Yeah, if you could include that additional bugfix either in an additional commit here or in a separate PR, that would be great!

**Comment by MachX428** [comment]
:rofl:
Ok, will do. I'll add it as a separate PR since its kind of its own thing. I also looked into it a little more today and now understand it a lot more, and do think it will resolve issue #17 (even though I can't full verify that part of it myself since I'm on Klipper right now).
